openstack/03a-IPHC/frag.c

```diff
@@ -0,0 +1,756 @@
+
+            // copy address information
+            lowpan_fragment->l3_destinationAdd = msg->l3_destinationAdd;
+            lowpan_fragment->l3_sourceAdd = msg->l3_sourceAdd;
+            lowpan_fragment->l2_nextORpreviousHop = msg->l2_nextORpreviousHop;
+
+            remaining_bytes -= fragment_length;
+
+            if (fragment_offset == 0) {
+                prepend_frag1_header(lowpan_fragment, msg->length, frag_vars.global_tag);
+            } else {
+                prepend_fragn_header(lowpan_fragment, msg->length, frag_vars.global_tag, fragment_offset);
+            }
+
+            // update the fragment offset
+            fragment_offset += (fragment_length / OFFSET_MULTIPLE);
+        }
+
+        // send all the fragments with the current datagram tag
+        for (int i = 0; i < FRAGMENT_BUFFER_SIZE; i++) {
+            if (frag_vars.fragmentBuf[i].datagram_tag == frag_vars.global_tag) {
+                // try to send the fragment. If this fails, abort the transmission of the other fragments.
+                if (sixtop_send(frag_vars.fragmentBuf[i].pFragment) == E_FAIL) {
+                    openserial_printError(COMPONENT_FRAG, ERR_FRAG_TX_FAIL,
+                                          (errorparameter_t) frag_vars.global_tag,
+                                          (errorparameter_t) frag_vars.fragmentBuf[i].datagram_offset);
+                    cleanup_fragments(frag_vars.global_tag);
+                    return E_FAIL;
+                } else {
+                    // fragment succesfully scheduled, lock it
+                    LOCK(frag_vars.fragmentBuf[i]);
+                }
+            }
+        }
+
+        // if we arrive here, all fragments were successfully created and passed to the MAC layer
+        return E_SUCCESS;
+
+    } else if (msg->l3_isFragment) {
+        // this is a fragment (type frag1) that's being source routed
+        // set nexthop in vrb and restore original frag1 header
+
+        for (uint8_t i = 0; i < NUM_OF_VRBS; i++) {
+            if (frag_vars.vrbs[i].frag1 == msg) {
+                memcpy(&frag_vars.vrbs[i].nexthop, &msg->l2_nextORpreviousHop, sizeof(open_addr_t));
+                prepend_frag1_header(msg, frag_vars.vrbs[i].size, frag_vars.vrbs[i].tag);
+                fast_forward_frags(frag_vars.vrbs[i].tag, frag_vars.vrbs[i].size, i);
+                break;
+            }
+        }
+
+        // forward source routed fragment
+        return sixtop_send(msg);
+    } else {
+        msg->l3_isFragment = FALSE;
+        return sixtop_send(msg);
+    }
+}
+
+
+void frag_sendDone(OpenQueueEntry_t *msg, owerror_t sendError) {
+    bool frags_queued;
+    bool upward_relay;
+    uint16_t datagram_tag;
+    OpenQueueEntry_t *original_msg;
+
+    if (msg->l3_isFragment && !msg->l3_useSourceRouting) {
+
+        frags_queued = FALSE;
+        upward_relay = FALSE;
+
+        // delete sent fragment from fragment buffer
+        int i;
```

Comment:
Usually, I define all various at the beginning of the function. Merely a C99 style.
Bit, this is trivle.